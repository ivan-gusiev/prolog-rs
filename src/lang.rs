use std::fmt::{Display, Formatter};
use std::iter::once;

use symbol::{SymDisplay, Symbol};

use crate::symbol::SymbolTable;

pub type VarName = Symbol;
pub type FunctorName = Symbol;
pub type Arity = u32;

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
pub struct Functor(pub FunctorName, pub Arity);

impl Functor {
    pub fn name(&self) -> VarName {
        self.0
    }

    pub fn arity(&self) -> Arity {
        self.1
    }
}

impl Display for Functor {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        write!(f, "{}/{}", self.0, self.1)
    }
}

impl SymDisplay for Functor {
    fn sym_fmt(
        &self,
        f: &mut Formatter<'_>,
        symbol_table: &SymbolTable,
    ) -> Result<(), std::fmt::Error> {
        self.0.sym_fmt(f, symbol_table)?;
        write!(f, "/{}", self.1)
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct Struct(Functor, Vec<Term>);

impl Struct {
    pub fn new(functor: Functor, terms: &[Term]) -> Result<Struct, String> {
        if terms.len() != functor.arity() as usize {
            return Err(format!(
                "functor {functor} arity does not correspond to the terms count"
            ));
        }

        Ok(Struct(functor, terms.to_vec()))
    }

    pub fn from_name(name: FunctorName, terms: &[Term]) -> Struct {
        // unwrap is fine because we 100% know that the number of terms and arity matches
        Self::new(Functor(name, terms.len() as u32), terms).unwrap()
    }

    pub fn constant(name: FunctorName) -> Struct {
        Self::new(Functor(name, 0), &[]).unwrap()
    }

    pub fn functor(&self) -> Functor {
        self.0
    }

    pub fn terms(&self) -> &[Term] {
        self.1.as_slice()
    }

    pub fn is_const(&self) -> bool {
        self.functor().arity() == 0
    }
}

impl Display for Struct {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        write!(f, "{}", self.functor().name())?;

        if !self.is_const() {
            write!(f, "({}", self.terms()[0])?;
            for term in self.terms().iter().skip(1) {
                write!(f, ", {term}")?
            }
            write!(f, ")")?
        }

        Ok(())
    }
}

impl SymDisplay for Struct {
    fn sym_fmt(
        &self,
        f: &mut Formatter<'_>,
        symbol_table: &SymbolTable,
    ) -> Result<(), std::fmt::Error> {
        self.functor().name().sym_fmt(f, symbol_table)?;

        if !self.is_const() {
            write!(f, "(")?;
            self.terms()[0].sym_fmt(f, symbol_table)?;
            for term in self.terms().iter().skip(1) {
                write!(f, ", ")?;
                term.sym_fmt(f, symbol_table)?;
            }
            write!(f, ")")?
        }

        Ok(())
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub enum Term {
    Variable(VarName),
    Struct(Struct),
}

impl Display for Term {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        match self {
            Self::Variable(nm) => write!(f, "{nm}"),
            Self::Struct(s) => s.fmt(f),
        }
    }
}

impl SymDisplay for Term {
    fn sym_fmt(
        &self,
        f: &mut Formatter<'_>,
        symbol_table: &SymbolTable,
    ) -> Result<(), std::fmt::Error> {
        match self {
            Self::Variable(nm) => nm.sym_fmt(f, symbol_table),
            Self::Struct(s) => s.sym_fmt(f, symbol_table),
        }
    }
}

impl Term {
    pub fn into_struct(self) -> Option<Struct> {
        match self {
            Self::Struct(s) => Some(s),
            _ => None,
        }
    }
}

peg::parser!(
    grammar prolog_parser() for str {
        rule whitespace()
            = quiet!{[' ' | '\t' | '\r' | '\n']}

        rule line_comment()
            = quiet!{("#" / "%") [^'\n']*} / expected!("line comment")

        rule block_comment()
            = quiet!{"/*" (!"*/" [_])* "*/"} / expected!("block comment")

        rule comment()
            = line_comment() / block_comment()

        rule rest() -> Vec<char>
            = ['a'..='z' | 'A'..='Z' | '0'..='9' | '_']*

        rule varname() -> std::iter::Chain<std::iter::Once<char>, std::vec::IntoIter<char>>
            = quiet!{ initial:['A'..='Z'] rest:rest() {once(initial).chain(rest.into_iter())}}
            / expected!("variable name")

        rule structname() -> std::iter::Chain<std::iter::Once<char>, std::vec::IntoIter<char>>
            = quiet!{ initial:['a'..='z'] rest:rest() {once(initial).chain(rest.into_iter())}}
            / expected!("structure name")

        rule _()
            = whitespace()* comment()?

        rule variable(symbols: &mut SymbolTable) -> Term
            = n:varname() { Term::Variable(symbols.intern_chars( n )) }

        rule structure(symbols: &mut SymbolTable) -> Term
            = n:structname() "(" ts:(term(symbols) ** ",") ")" { Term::Struct(Struct::from_name(symbols.intern_chars(n), ts.as_slice())) }

        rule constant(symbols: &mut SymbolTable) -> Term
            = n:structname() { Term::Struct(Struct::constant(symbols.intern_chars(n))) }

        pub rule term(symbols: &mut SymbolTable) -> Term
            = _ t:(variable(symbols) / structure(symbols) / constant(symbols)) _ { t }
    }
);

pub fn parse_term(term: &str, symbol_table: &mut SymbolTable) -> Result<Term, String> {
    prolog_parser::term(term, symbol_table).map_err(|e| format!("{e}"))
}

pub fn parse_struct(term: &str, symbol_table: &mut SymbolTable) -> Result<Struct, String> {
    parse_term(term, symbol_table)?
        .into_struct()
        .ok_or("term must be a struct.".to_string())
}

#[cfg(test)]
mod unittests {
    use std::convert::TryFrom;

    use crate::symbol::Symbol;

    use super::{Functor, Struct, Term};

    fn dirty(x: char) -> Symbol {
        Symbol::try_from(x.to_string()).unwrap()
    }

    #[test]
    fn struct_new_success() {
        assert!(matches!(
            Struct::new(
                Functor(dirty('f'), 1),
                &[Term::Struct(Struct::constant(dirty('h')))]
            ),
            Ok(_)
        ))
    }

    #[test]
    fn struct_new_fail() {
        assert!(matches!(
            Struct::new(
                Functor(dirty('f'), 2),
                &[Term::Struct(Struct::constant(dirty('h')))]
            ),
            Err(_)
        ))
    }
}
