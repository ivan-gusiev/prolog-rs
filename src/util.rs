use std::{
    collections::HashMap,
    fmt::{Debug, Display, Error, Write},
};

use crate::{
    asm::Assembly,
    compile::CompileInfo,
    data::{Addr, CodePtr, RegPtr},
    instr::Instruction,
    lang::{Functor, Struct},
    machine::{Machine, MachineError},
    symbol::{to_display, SymDisplay, SymbolTable},
    var::VarBindings,
};

pub fn collapse<T>(result: Result<T, T>) -> T {
    match result {
        Ok(x) => x,
        Err(x) => x,
    }
}

#[must_use]
pub fn writeout<T: Display, I: IntoIterator<Item = T>>(items: I) -> String {
    fn writeout_impl<T: Display, I: IntoIterator<Item = T>>(items: I) -> Result<String, Error> {
        let mut out = String::new();
        for (idx, item) in items.into_iter().enumerate() {
            writeln!(out, "{idx:#03}\t{item}")?;
        }
        Ok(out)
    }

    writeout_impl(items).unwrap_or_else(|e| format!("{e}"))
}

pub fn writeout_table2<T: Display, U: Display>(ts: &[T], us: &[U]) -> String {
    let items = ts.iter().zip(us.iter()).map(|(t, u)| format!("{t}\t{u}"));
    writeout(items)
}

pub fn writeout_dict<T: Display + Ord, U: Display, D>(dict: D) -> String
where
    D: IntoIterator<Item = (T, U)>,
{
    let mut items_vec = dict.into_iter().collect::<Vec<_>>();
    items_vec.sort_by(|(a, _), (b, _)| a.cmp(b));
    let items = items_vec.into_iter().map(|(t, u)| format!("{t}\t{u}"));
    writeout(items)
}

pub fn writeout_sym<T: SymDisplay>(items: &[T], symbol_table: &SymbolTable) -> String {
    writeout(items.iter().map(|item| to_display(item, symbol_table)))
}

pub fn writeout_table2_sym<T: SymDisplay, U: SymDisplay>(
    ts: &[T],
    us: &[U],
    symbol_table: &SymbolTable,
) -> String {
    let items = ts.iter().zip(us.iter()).map(|(t, u)| {
        format!(
            "{}\t{}",
            to_display(t, symbol_table),
            to_display(u, symbol_table)
        )
    });
    writeout(items)
}

pub fn case<T: Display, U: Display>(input: T, output: U) -> String {
    format!("{input}\n-----\n{output}")
}

pub fn case_dbg<T: Debug, U: Debug>(input: T, output: U) -> String {
    format!("{input:#?}\n-----\n{output:#?}")
}

pub fn write_program_result(
    machine: &Machine,
    symbol_table: &mut SymbolTable,
    query_bindings: &VarBindings,
    maybe_program_bindings: Option<&VarBindings>,
) -> String {
    fn write_program_result_impl(
        machine: &Machine,
        symbol_table: &mut SymbolTable,
        query_bindings: &VarBindings,
        maybe_program_bindings: Option<&VarBindings>,
    ) -> Result<String, Error> {
        let mut out = String::new();
        writeln!(out, "HEAP ({})\n___", machine.heap_len())?;
        writeln!(out, "{}", writeout(machine.iter_heap().collect::<Vec<_>>()))?;
        writeln!(out, "QUERY\n----")?;
        writeln!(
            out,
            "{}",
            writeout_sym(
                &machine.describe_vars(query_bindings, symbol_table).unwrap(),
                symbol_table
            )
        )?;
        writeln!(out, "PROGRAM\n____")?;
        if let Some(program_bindings) = maybe_program_bindings {
            writeln!(
                out,
                "{}",
                writeout_sym(
                    &machine
                        .describe_vars(program_bindings, symbol_table)
                        .unwrap(),
                    symbol_table
                )
            )?;
        }
        Ok(out)
    }

    write_program_result_impl(
        machine,
        symbol_table,
        query_bindings,
        maybe_program_bindings,
    )
    .unwrap()
}

pub fn writeout_annotated_mappings(
    machine: &Machine,
    query_bindings: &VarBindings,
    program_bindings: &VarBindings,
    symbol_table: &SymbolTable,
) -> String {
    fn writeout_annotated_mappings_impl(
        machine: &Machine,
        query_bindings: &VarBindings,
        program_bindings: &VarBindings,
        symbol_table: &SymbolTable,
    ) -> Result<String, Error> {
        let mut out = String::new();
        let mut vars = Vec::<String>::with_capacity(2);
        for reg_id in 1..machine.iter_reg().len() {
            let reg = RegPtr(reg_id);
            vars.clear();
            if let Some(var) = query_bindings.get(&machine.trace_heap(reg.into()).unwrap()) {
                vars.push(format!("query.{}", var.sym_to_str(symbol_table)))
            }
            if let Some(var) = program_bindings.get(&machine.trace_heap(reg.into()).unwrap()) {
                vars.push(format!("program.{}", var.sym_to_str(symbol_table)))
            }
            let annotations = if vars.is_empty() {
                String::new()
            } else {
                format!("// {}", vars.join(", "))
            };
            writeln!(out, "{reg}\t{annotations}")?
        }
        Ok(out)
    }

    writeout_annotated_mappings_impl(machine, query_bindings, program_bindings, symbol_table)
        .unwrap_or_else(|e| format!("{e}"))
}

pub fn writeout_compile_result(compile_result: &CompileInfo, symbol_table: &SymbolTable) -> String {
    fn writeout_compile_result_impl(
        compile_result: &CompileInfo,
        symbol_table: &SymbolTable,
    ) -> Vec<String> {
        let mut out = Vec::<String>::new();
        let mut annotations = Vec::<String>::with_capacity(2);
        let process_local = |addr: &Addr, ann: &mut Vec<String>| {
            if let Some(local) = addr.get_local() {
                if let Some(var) = compile_result.var_mapping.get(&local) {
                    ann.push(format!("{}={}", local, to_display(&var, symbol_table)))
                }
            }
        };
        let process_locals = |rs: &[&Addr], ann: &mut Vec<String>| {
            for r in rs {
                process_local(r, ann)
            }
        };

        for warning in compile_result.warnings.iter() {
            out.push(format!("; WARN: {}", to_display(warning, symbol_table)));
        }

        for instr in compile_result.instructions.iter() {
            annotations.clear();
            match instr {
                Instruction::GetStructure(_, l)
                | Instruction::PutStructure(_, l)
                | Instruction::UnifyValue(l)
                | Instruction::UnifyVariable(l)
                | Instruction::SetValue(l)
                | Instruction::SetVariable(l) => process_local(l, &mut annotations),
                Instruction::GetValue(l1, l2)
                | Instruction::GetVariable(l1, l2)
                | Instruction::PutValue(l1, l2)
                | Instruction::PutVariable(l1, l2) => process_locals(&[l1, l2], &mut annotations),
                _ => (),
            }
            let comment = if annotations.is_empty() {
                String::new()
            } else {
                format!("// {}", annotations.join(", "))
            };
            out.push(format!("{}\t{}", to_display(instr, symbol_table), comment))
        }
        out
    }

    writeout(
        writeout_compile_result_impl(compile_result, symbol_table)
            .iter()
            .map(String::as_str)
            .map(str::trim),
    )
}

pub fn writeout_assembly(assembly: &Assembly, symbol_table: &SymbolTable) -> String {
    let mut ptr_to_label = HashMap::<usize, Vec<Functor>>::with_capacity(assembly.label_map.len());
    for (lbl, ptr) in assembly.label_map.iter() {
        let vec = ptr_to_label.entry(ptr.0).or_insert_with(Vec::new);
        match vec.binary_search(lbl) {
            Ok(_) => {} // element already in vector @ `pos`
            Err(pos) => vec.insert(pos, *lbl),
        }
    }

    let entrypoint_ptr = assembly
        .entry_point
        .as_ref()
        .map(|e| e.location.0)
        .unwrap_or(usize::MAX);
    let lines = assembly.instructions.iter().enumerate().map(|(i, instr)| {
        let labels = ptr_to_label.get(&i).map(Vec::as_slice).unwrap_or(&[]);

        let write_labels = WriteVec::new(labels)
            .with_separator(": ")
            .with_trailing_separator(true);
        format!(
            "{}{}{}",
            if entrypoint_ptr == i {
                "ENTRYPOINT: "
            } else {
                ""
            },
            to_display(&write_labels, symbol_table),
            to_display(instr, symbol_table)
        )
    });

    writeout(lines)
}

pub struct WriteVec<'a, T> {
    data: &'a [T],
    separator: &'a str,
    opener: Option<&'a str>,
    closer: Option<&'a str>,
    trailing_separator: bool,
}

impl<'a, T> WriteVec<'a, T> {
    pub fn new(data: &'a [T]) -> Self {
        Self {
            data,
            separator: ", ",
            opener: None,
            closer: None,
            trailing_separator: false,
        }
    }

    pub fn with_separator(self, separator: &'a str) -> Self {
        Self { separator, ..self }
    }

    pub fn with_brackets(self, opener: Option<&'a str>, closer: Option<&'a str>) -> Self {
        Self {
            opener,
            closer,
            ..self
        }
    }

    pub fn with_trailing_separator(self, trailing_separator: bool) -> Self {
        Self {
            trailing_separator,
            ..self
        }
    }

    pub fn with_square_brackets(self) -> Self {
        self.with_brackets(Some("["), Some("]"))
    }
}

impl<'a, T: SymDisplay> SymDisplay for WriteVec<'a, T> {
    fn sym_fmt(
        &self,
        f: &mut std::fmt::Formatter<'_>,
        symbol_table: &SymbolTable,
    ) -> Result<(), Error> {
        if let Some(opener) = self.opener {
            write!(f, "{opener}")?;
        }
        if !self.data.is_empty() {
            write!(f, "{}", to_display(&self.data[0], symbol_table))?;
            for item in &self.data[1..] {
                write!(f, "{}{}", self.separator, to_display(item, symbol_table))?;
            }
            if self.trailing_separator {
                write!(f, "{}", self.separator)?;
            }
        }
        if let Some(closer) = self.closer {
            write!(f, "{closer}")?;
        }
        Ok(())
    }
}

pub fn run_just_query(
    machine: &mut Machine,
    instructions: &[Instruction],
) -> Result<(), MachineError> {
    machine.set_code(&[Instruction::Proceed]);
    let p = machine.append_code(instructions);
    machine.set_p(p);
    machine.set_cp(CodePtr(instructions.len()) + 1);
    machine.execute().run()
}

pub fn lbl_for(goals: &[Struct]) -> HashMap<Functor, CodePtr> {
    goals
        .iter()
        .map(|goal| (goal.functor(), CodePtr(0)))
        .collect()
}
