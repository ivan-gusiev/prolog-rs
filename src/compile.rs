use std::collections::{HashMap, HashSet, VecDeque};
use std::fmt::{Display, Formatter};
use std::hash::Hash;

use instr::Instruction;
use lang::{Term, VarName};

// TODO: for some reason rustc requires `extern crate` definitions, fix this
extern crate topological_sort;
use self::topological_sort::TopologicalSort;

use crate::{data::RegPtr, lang::Functor};

#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
struct TermId(usize);

impl TermId {
    fn inc(&mut self) {
        self.0 += 1;
    }
}

impl Display for TermId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "t{}", self.0)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
enum FlatRef {
    Term(TermId), // TODO: this shouldn't exist after flatten_term returns
    Register(RegPtr),
}

impl Display for FlatRef {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            FlatRef::Term(id) => id.fmt(f),
            FlatRef::Register(ptr) => ptr.fmt(f),
        }
    }
}

#[derive(Debug, Hash, PartialEq, Eq)]
struct FlatStruct(Functor, Vec<FlatRef>);

impl Display for FlatStruct {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}(", self.0)?;
        for term in self.1.iter() {
            write!(f, "{term},")?
        }
        write!(f, ")")
    }
}

#[derive(Debug, Hash, PartialEq, Eq)]
enum FlattenedTerm {
    Variable(VarName),
    Struct(FlatStruct),
}

impl Display for FlattenedTerm {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            FlattenedTerm::Variable(n) => n.fmt(f),
            FlattenedTerm::Struct(s) => s.fmt(f),
        }
    }
}

impl FlattenedTerm {
    fn new_str(f: Functor, refs: Vec<FlatRef>) -> FlattenedTerm {
        FlattenedTerm::Struct(FlatStruct(f, refs))
    }

    fn get_str(&self) -> &FlatStruct {
        if let FlattenedTerm::Struct(str) = self {
            str
        } else {
            panic!("not a struct")
        }
    }
}

fn flatten_query(query: Term) -> (Vec<FlattenedTerm>, HashMap<VarName, RegPtr>) {
    let mut term_map = HashMap::<TermId, RegPtr>::new();
    let mut var_map = HashMap::<VarName, RegPtr>::new();
    let mut queue = VecDeque::from([(&query, TermId(0))]);
    let mut result = Vec::<FlattenedTerm>::new();
    let mut term_counter = TermId(0);

    let mut next_term_id = || {
        term_counter.inc();
        term_counter
    };

    loop {
        match queue.pop_front() {
            None => break,
            Some((Term::Variable(nm), id)) => {
                if var_map.contains_key(nm) {
                    term_map.insert(id, var_map[nm]);
                } else {
                    result.push(FlattenedTerm::Variable(*nm));
                    var_map.insert(*nm, RegPtr(result.len()));
                    term_map.insert(id, RegPtr(result.len()));
                }
            }
            Some((Term::Struct(str), id)) => {
                let mut subterms: Vec<FlatRef> = vec![];

                for subterm in str.terms() {
                    match subterm {
                        v @ Term::Variable(nm) => {
                            if var_map.contains_key(nm) {
                                subterms.push(FlatRef::Register(var_map[nm]))
                            } else {
                                let id = next_term_id();
                                subterms.push(FlatRef::Term(id));
                                queue.push_back((v, id));
                            }
                        }
                        t @ Term::Struct(_) => {
                            let id = next_term_id();
                            subterms.push(FlatRef::Term(id));
                            queue.push_back((t, id));
                        }
                    }
                }
                result.push(FlattenedTerm::new_str(str.functor(), subterms));
                term_map.insert(id, RegPtr(result.len()));
            }
        }
    }

    for term in result.iter_mut() {
        match term {
            FlattenedTerm::Variable(_) => (),
            FlattenedTerm::Struct(FlatStruct(_, subterms)) => {
                for subterm in subterms {
                    if let FlatRef::Term(id) = subterm {
                        match term_map.get(id) {
                            Some(reg) => *subterm = FlatRef::Register(*reg),
                            None => panic!("Could not find a term for {:?}", id),
                        }
                    }
                }
            }
        }
    }

    (result, var_map)
}

fn ptoi(ptr: RegPtr) -> usize {
    ptr.0 - 1
}

fn itop(idx: usize) -> RegPtr {
    RegPtr(idx + 1)
}

fn extract_structs(terms: &[FlattenedTerm]) -> Vec<RegPtr> {
    let mut structs = vec![];
    for (i, term) in terms.iter().enumerate() {
        match term {
            FlattenedTerm::Struct(_) => structs.push(itop(i)),
            FlattenedTerm::Variable(_) => (),
        }
    }
    structs
}

fn order_query_structs(terms: &[FlattenedTerm], structs_to_sort: &[RegPtr]) -> Vec<RegPtr> {
    fn regs(term: &FlattenedTerm) -> HashSet<RegPtr> {
        let mut result = HashSet::new();

        if let FlattenedTerm::Struct(FlatStruct(_, refs)) = term {
            for r in refs {
                if let FlatRef::Register(ptr) = r {
                    result.insert(*ptr);
                }
            }
        }

        result
    }

    let mut ts = TopologicalSort::<RegPtr>::new();
    for l in 0..structs_to_sort.len() {
        for r in (l + 1)..structs_to_sort.len() {
            let lp = structs_to_sort[l];
            let rp = structs_to_sort[r];
            let lhs = &terms[ptoi(lp)];
            let rhs = &terms[ptoi(rp)];

            let regs_lhs = regs(lhs);
            let regs_rhs = regs(rhs);

            if regs_lhs.contains(&rp) {
                ts.add_dependency(rp, lp);
            } else if regs_rhs.contains(&lp) {
                ts.add_dependency(lp, rp);
            }
        }
    }

    let mut result = vec![];
    loop {
        let mut batch = ts.pop_all(); // use pop_all to preserve the original ordering
        if batch.is_empty() {
            break;
        }

        batch.sort();
        result.append(&mut batch)
    }

    // special case when there are no dependencies between terms, e.g. only one term to sort
    if result.is_empty() && structs_to_sort.len() == 1 {
        result.push(structs_to_sort[0]);
    }

    result
}

pub fn compile_query(query: Term) -> CompileResult {
    let (registers, vars) = flatten_query(query);
    let structs = order_query_structs(&registers, &extract_structs(&registers));
    let mut seen = HashSet::<RegPtr>::new();
    let mut result = vec![];

    for struct_ptr in structs {
        let FlatStruct(f, refs) = registers[ptoi(struct_ptr)].get_str();
        result.push(Instruction::PutStructure(*f, struct_ptr));
        seen.insert(struct_ptr);
        for flat_ref in refs {
            if let FlatRef::Register(ref_ptr) = flat_ref {
                if seen.contains(ref_ptr) {
                    result.push(Instruction::SetValue(*ref_ptr))
                } else {
                    seen.insert(*ref_ptr);
                    result.push(Instruction::SetVariable(*ref_ptr))
                }
            }
        }
    }

    CompileResult {
        instructions: result,
        var_mapping: vars.into(),
    }
}

pub fn compile_program(program: Term) -> CompileResult {
    let (registers, vars) = flatten_query(program);
    let structs = extract_structs(&registers);
    let mut seen = HashSet::<RegPtr>::new();
    let mut result = vec![];

    for struct_ptr in structs {
        let FlatStruct(f, refs) = registers[ptoi(struct_ptr)].get_str();
        result.push(Instruction::GetStructure(*f, struct_ptr));
        seen.insert(struct_ptr);
        for flat_ref in refs {
            if let FlatRef::Register(ref_ptr) = flat_ref {
                if seen.contains(ref_ptr) {
                    result.push(Instruction::UnifyValue(*ref_ptr))
                } else {
                    seen.insert(*ref_ptr);
                    result.push(Instruction::UnifyVariable(*ref_ptr))
                }
            }
        }
    }

    CompileResult {
        instructions: result,
        var_mapping: vars.into(),
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct VarMapping(HashMap<RegPtr, VarName>);

impl From<HashMap<VarName, RegPtr>> for VarMapping {
    fn from(value: HashMap<VarName, RegPtr>) -> Self {
        let mut result = HashMap::<RegPtr, VarName>::new();
        for (v, r) in value.into_iter() {
            result.insert(r, v);
        }
        Self(result)
    }
}

impl VarMapping {
    pub fn mappings(&self) -> impl Iterator<Item = (&VarName, &RegPtr)> {
        self.0.iter().map(|(r, v)| (v, r))
    }

    pub fn get(&self, ptr: &RegPtr) -> Option<VarName> {
        self.0.get(ptr).copied()
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct CompileResult {
    pub instructions: Vec<Instruction>,
    pub var_mapping: VarMapping,
}

#[cfg(test)]
use crate::symbol::SymbolTable;

#[test]
fn test_order_structs() {
    use lang::parse_term;
    let mut symbol_table = SymbolTable::new();
    let query = parse_term("p(Z,h(Z,W),f(W))", &mut symbol_table).unwrap();
    let (terms, _) = flatten_query(query);
    let structs = extract_structs(&terms);

    assert_eq!(
        &order_query_structs(&terms, &structs),
        &[RegPtr(3), RegPtr(4), RegPtr(1)]
    );
}

#[test]
fn test_flatten_query() {
    use lang::parse_term;
    let mut symbol_table = SymbolTable::new();
    let query = parse_term("p(Z,h(Z,W),f(W))", &mut symbol_table).unwrap();

    let results = flatten_query(query)
        .0
        .into_iter()
        .map(|x| format!("{x:?}"))
        .collect::<Vec<_>>()
        .join("||\n");

    assert_eq!(
        format!("{results}"),
        r#"Struct(FlatStruct(Functor(:p, 3), [Register(RegPtr(2)), Register(RegPtr(3)), Register(RegPtr(4))]))||
Variable(:Z)||
Struct(FlatStruct(Functor(:h, 2), [Register(RegPtr(2)), Register(RegPtr(5))]))||
Struct(FlatStruct(Functor(:f, 1), [Register(RegPtr(5))]))||
Variable(:W)"#.to_string()
    );
}

#[test]
fn test_compile_query() {
    use lang::parse_term;
    let mut symbol_table = SymbolTable::new();
    let query = parse_term("p(Z,h(Z,W),f(W))", &mut symbol_table).unwrap();
    let instructions = Instruction::from_program(
        r#"
        put_structure h/2, X3
        set_variable X2      
        set_variable X5      
        put_structure f/1, X4
        set_value X5         
        put_structure p/3, X1
        set_value X2         
        set_value X3         
        set_value X4         
        "#,
        &mut symbol_table,
    )
    .unwrap();
    assert_eq!(
        compile_query(query),
        CompileResult {
            instructions,
            var_mapping: HashMap::from([
                (symbol_table.intern("W"), RegPtr(5)),
                (symbol_table.intern("Z"), RegPtr(2))
            ])
            .into()
        }
    )
}

#[test]
fn test_compile_program() {
    use lang::parse_term;
    let mut symbol_table = SymbolTable::new();
    let program = parse_term("p(f(X), h(Y,f(a)), Y)", &mut symbol_table).unwrap();
    let instructions = Instruction::from_program(
        r#"
        get_structure p/3, X1
        unify_variable X2
        unify_variable X3
        unify_variable X4
        get_structure f/1, X2
        unify_variable X5
        get_structure h/2, X3
        unify_value X4
        unify_variable X6
        get_structure f/1, X6
        unify_variable X7
        get_structure a/0, X7
        "#,
        &mut symbol_table,
    )
    .unwrap();
    assert_eq!(
        compile_program(program),
        CompileResult {
            instructions,
            var_mapping: HashMap::from([
                (symbol_table.intern("X"), RegPtr(5)),
                (symbol_table.intern("Y"), RegPtr(4))
            ])
            .into()
        }
    )
}
