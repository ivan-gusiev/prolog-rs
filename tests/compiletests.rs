extern crate insta;
extern crate parameterized;
extern crate prolog_rs;

#[cfg(test)]
mod compiletests {
    use insta::assert_display_snapshot;
    use parameterized::parameterized;
    use prolog_rs::compile::{compile_program, compile_query};
    use prolog_rs::lang::parse_term;
    use prolog_rs::util::{case, writeout};

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "a"
    })]
    fn test_query_compile(input: &str) {
        let query = parse_term(input).unwrap();
        let instructions = compile_query(query);

        assert_display_snapshot!(case(input, writeout(&instructions)));
    }

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "p(f(X), h(Y, f(a)), Y)",
        "a"
    })]
    fn test_program_compile(input: &str) {
        let query = parse_term(input).unwrap();
        let instructions = compile_program(query);

        assert_display_snapshot!(case(input, writeout(&instructions)));
    }
}
