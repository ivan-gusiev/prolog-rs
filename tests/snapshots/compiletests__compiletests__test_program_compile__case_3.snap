---
source: tests/compiletests.rs
expression: "case(input, writeout(&instructions))"
---
f(b, Y)
-----
000	get_structure f/2, X1
001	unify_variable X2
002	unify_variable X3
003	get_structure b/0, X2

